### PR TITLE
Add image export options

### PR DESCRIPTION
**Pull request by danfishgold**
Add the option to export the annotated code as an image (png), either as a download or pasted into the clipboard.

Screen Shot 2021-08-05 at 19 21 19.

This change should make it much easier to share annotations in a Miro board during code reading clubs.

The clipboard API doesn't allow jpeg images for some reason. I figured jpeg artifacts might be a problem with images that might be very big with very small text, so that constraint might be a good thing.

Also worth noting: [`Clipboard.write()`](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/write) is not supported yet on Firefox, so the copy button is disabled there (via a check of `navigator.clipboard.write !== undefined`).

Both the download and copy options use [html2canvas](http://html2canvas.hertzen.com) to generate the image. There's a problem where it ignores the `mix-blend-mode` css rule, which is _very_ important for annotations, so I had to use [patch-package](https://www.npmjs.com/package/patch-package) to patch that in (I used the fix suggested in [this issue](https://github.com/niklasvh/html2canvas/issues/580#issuecomment-635450979)).

Exporting to pdf is probably very useful but it can't be done on the client side (other than printing to pdf). In order to solve this problem we'd have to do something server-side. This might be a good opportunity to combine this project and [CodeReadingClibs/pdf-maker](https://github.com/CodeReadingClubs/pdf-maker).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,7 +27,6 @@ Netlify.
   near the top/bottom)
 - Footer with link to github and a "try your own code" button
 - Share links (needs hosting :((( )
-- Download/save/copy as image/pdf
```
Comment (katjam) [inline README.md:30]: :tada:

File: package.json [hunk 2]
```diff
@@ -4,11 +4,14 @@
   "scripts": {
     "dev": "vite",
     "build": "tsc && vite build",
-    "serve": "vite preview"
+    "serve": "vite preview",
+    "postinstall": "patch-package"
```
Comment (katjam) [inline package.json:8]: That's cool. I've not seen patches applied this way before.

File: src/hooks/useExport.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
+import html2canvas from 'html2canvas'
+import { useCallback, useState } from 'react'
+
+export type CopyState = 'idle' | 'preparing' | 'success' | 'failure'
+
+export type DownloadState = 'idle' | 'preparing' | 'failure'
```
Comment (katjam) [inline src/hooks/useExport.ts:6]: I guess it makes sense not having success state here
vs
Needing a success state on copy so there is a noticeable side-effect.

File: src/components/Export.tsx [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      return 'copy as png'
+    }
+    case 'success': {
+      return 'copied!'
+    }
+  }
+}
+
+function downloadButtonTitle(downloadState: DownloadState): string {
+  switch (downloadState) {
```
Comment (katjam) [inline src/components/Export.tsx:45]: Loving the type safe pattern matches.

**Review by katjam — APPROVED** [review]
This all looks pretty sensible to me. Thanks for the doc on the PR explaining your choices too.

File: package.json (hunk 2 above)
Comment (danfishgold) [reply package.json:8]: I saw it mentioned on twitter a couple of months ago and I was just waiting for the opportunity to use it 😄

File: src/hooks/useExport.ts (hunk 3 above)
Comment (danfishgold) [reply src/hooks/useExport.ts:6]: yeah, exactly! there's a one second timer that changes the button text to "copied"

File: src/components/Export.tsx (hunk 4 above)
Comment (danfishgold) [reply src/components/Export.tsx:45]: you love to see it

**Comment by danfishgold** [comment]
just one bug fix with regard to differences in `ClipboardItem` implementations between Chromium and WebKit and it's good to go

**Pull request merged**